TBB DOC : Dev Guide: Task Scheduler Bypass and How Does Task Scheduler Works (#521)
oneapi-src/oneTBB: doc/main/tbb_userguide/Task_Scheduler_Bypass.rst

.. _Task_Scheduler_Bypass:

Task Scheduler Bypass
=====================

Scheduler bypass is an optimization where you directly specify the next task to run. 
By the execution rules in  :doc:`How Task Scheduling Works <How_Does_Task_Scheduler_Works>`, 
spawning of the new task causes the executing thread to do the following:

 -  Push new task onto the thread's deque.
 -  Continue executing current task till completion.
 -  Pop the task from the thread's deque, unless it is stolen by another thread.

> Steps 1 and 3 introduce unnecessary deque operations, or worse yet, permit stealing that can hurt 

Reviewer: ```suggestion
Steps 1 and 3 introduce unnecessary deque operations or, even worse, allow stealing that can hurt 
```

> locality without adding significant parallelism. These problems can be avoided by returning next task to execute 

Reviewer: ```suggestion
locality without adding significant parallelism. You can avoid these problems by returning the next task to execute 
```

instead of spawning it. When using the method shown in :doc:`How Task Scheduling Works <How_Does_Task_Scheduler_Works>`,

> returned task becomes the next task executed by the thread. Furthermore, this approach (almost) guarantees that 

Reviewer: ```suggestion
the returned task becomes the next task executed by the thread. Furthermore, this approach almost guarantees that 
```

> the task will be executed by the current thread, not some other thread.

Reviewer: ```suggestion
the task will be executed by the current thread, and not by any other thread.
```

.. _Task_Scheduler_Bypass:

Task Scheduler Bypass
=====================

Scheduler bypass is an optimization where you directly specify the next task to run. 
According to the rules of execution in  :doc:`How Task Scheduling Works <How_Does_Task_Scheduler_Works>`, 
spawning of the new task to be executed by the current thread, involves these steps:

 -  Push a new task onto the thread's deque.
 -  Continue to execute the current task until it is completed.
 -  Take a task from the thread's deque, unless it is stolen by another thread.

> Steps 1 and 3 introduce unnecessary deque operations or, even worse, allow stealing that can hurt 

Author: "even worse" should be left in place as cache disruption is the most serious drawback of the listed ones

> locality without adding significant parallelism. These problems can be avoided by returning next task to execute 

Reviewer: ```suggestion
locality without adding significant parallelism. These problems can be avoided by returning the next task to execute 
```

> locality without adding significant parallelism. These problems can be avoided by returning next task to execute 

Author: Need to mention "Bypass"

> locality without adding significant parallelism. These problems can be avoided by returning next task to execute 

Author: ```suggestion
locality without adding significant parallelism. These problems can be avoided by using "Task Scheduler Bypass" technique to directly point preferable task to execute 
```

Author: Does this looks better ?

> instead of spawning it. When using the method described in :doc:`How Task Scheduling Works <How_Does_Task_Scheduler_Works>`,

Author: ```suggestion
instead of spawning it. When, as described in :doc:`How Task Scheduling Works <How_Does_Task_Scheduler_Works>`,
```

> the returned task becomes the next task executed by the thread. Furthermore, this approach almost guarantees that 

Reviewer 2: The task does not always become the the next task due to Flow Graph priorities. Maybe `the returned task` is considered first to be executed by the thread`?

Author: good point, thanks for idea

Author: chnged

> the returned task becomes the next task executed by the thread. Furthermore, this approach almost guarantees that 

Author: ```suggestion
the returned task becomes the first candidate for the next task to be executed by the thread. Furthermore, this approach almost guarantees that 
```
